protocols/v2/subprotocols/common-messages/src/lib.rs

```diff
@@ -20,3 +27,77 @@ pub extern "C" fn _c_export_channel_endpoint_changed(_a: ChannelEndpointChanged)
 #[cfg(not(feature = "with_serde"))]
 #[no_mangle]
 pub extern "C" fn _c_export_setup_conn_succ(_a: SetupConnectionSuccess) {}
+
+#[cfg(feature = "prop_test")]
+impl ChannelEndpointChanged {
+    pub fn from_random(g: &mut Gen) -> Self {
+        ChannelEndpointChanged {
+            channel_id: u32::arbitrary(g).try_into().unwrap(),
+        }
+    }
+}
+
+#[cfg(feature = "prop_test")]
+impl SetupConnection<'static> {
+    pub fn from_random(g: &mut Gen) -> Self {
+        let protocol = setup_connection::Protocol::MiningProtocol;
+        // let protocol = setup_connection::Protocol::JobDistributionProtocol;
+        // let protocol = setup_connection::Protocol::TemplateDistributionProtocol;
+        // let protocol = setup_connection::Protocol::JobNegotiationProtocol;
+        let mut endpoint_host = Gen::new(255);
```

Comment:
Not sure if generate a new random generator inside `from_random`  is the right way to go or if the passed generator `g` should be used. Is not clear from quickcheck documentation we should ask somewhere which is the right pattern and put a comment with a reference to the answer.

Reply:
Is the below explanation accurate? If so, I will create an issue on quickcheck. If not, just lmk how I should modify it.

### Question: Idiomatic way to specify the length of an arbitrary vector


I have a struct with a field containing a data type that represents a vector of 255 bytes:
```
struct Foo {
  bar: B0255
}
```

Where

```
struct B0255<'b>(Inner<'b>);

enum Inner<'a> {
   Ref(&'a [u8]),
   Owned(Vec<u8>),
}
```

To test this struct, I have created a new struct and implemented `Arbitrary`, which calls a function `from_random` which is implemented for `Foo`.

```
struct RandomFoo(Foo);
impl Arbitrary from RandomFoo {
  fn arbitrary(g: &mut Gen) -> Self {
    RandomFoo(Foo::from_random(g))
  }
}
```

Currently, I am passing in the generator to the `from_random` function and generating the 255 byte vectors as:

```
impl Foo {
  fn from_random(g: &mut Gen) -> Self {
    let bar: B0255 = Vec<u8>::arbitrary(g).try_into().unwrap();
    Foo { bar }
  }
}
```


Alternatively, I can implement `from_random` as:

```
impl Foo {
  fn from_random(g: &mut Gen) -> Self {
    let mut bar: B0255 = Gen::new(255);
    bar: B0255 = Vec::<u8>::arbitrary(&mut bar).try_into().unwrap();
    Foo { bar }
  }
}
```

Which, if either, `from_random` implementation is the idiomatic way to generate a vector of a specific size?

Reply:
I would change the first example with:
```rust
impl Foo {
  fn from_random(g: &mut Gen) -> Self {
    
    let mut bar = Vec<u8>::arbitrary(g);
    bar.truncate(255);
    let bar: B0255 = bar.try_into().unwrap();
    Foo { bar }
  }
}
```
I truncate the vec to be sure to not panic with a too big vec. Testing if `try_into` return the right error when called with a too big value would be the best approach but not sure if it is possible as `arbitrary` need to return `Self` (I would ask also that in the issue) .

I would also add that:
1. we need `from_random` cause `Arbitrary` need to be implemented in another crate
2. `Arbitrary` need to be implemented for structs that can contains a lot different primitives (max bytes 64K, max bytes 255, exactly 32 bytes ecc ecc)
3. When `from_random` is implemented for fixed size primitives (eg U256) instead of `truncate` we use `resize(255, 0)` is that correct?

Reply:
https://github.com/BurntSushi/quickcheck/issues/297